envoyauth/response.go

```diff
@@ -343,54 +331,112 @@ func makeHeaderValueOption(k, v string) *ext_core_v3.HeaderValueOption {
 	}
 }
 
-func makeEnvoyHeaderValueOptionsFromHeadersMap(hvo []*ext_core_v3.HeaderValueOption, headers map[string]any) ([]*ext_core_v3.HeaderValueOption, error) {
-	hvo = slices.Grow(hvo, len(headers))
+func (result *EvalResult) getHeadersFromDecision(fieldName string) ([]*ext_core_v3.HeaderValueOption, error) {
+	return getHeadersWithTransformation(result.Decision, fieldName, preallocateForEnvoyHeaders, collectEnvoyHeaders)
+}
+
+func (result *EvalResult) getHTTPHeadersFromDecision(fieldName string) (http.Header, error) {
+	return getHeadersWithTransformation(result.Decision, fieldName, preallocateForHTTPHeaders, collectHTTPHeaders)
+}
+
+func getHeadersWithTransformation[T any](
```

Comment:
IMHO, a new approach looks a bit complicated and not easy to follow the logic.

Could you make just a function which converts the result from `getHeadersFromDecision` into `http.Header`?

Reply:
It's is an effort to balance complexity and performance for both cases of returning `HeaderValueOption` and `http.Header`. 
This suggested approach introduces an unnecessary intermediate conversion to `ext_core_v3.HeaderValueOption` for http.Header returning case affecting performance in that scenario. 
This concern directs us towards the implemented approach which keep both the cases optimized for highload services.

Reply:
Could you introduce the benchmark of `getHeadersFromDecision` before and after your refactoring?

Reply:
Moreover, you already made an intermediate conversion using `extractHeadersFromDecision` function.

Reply:
@regeda I respect your requirements, but it's probably better to follow up with PRs of your own to improve performance where needed rather than making requests for that in others contributions.

Reply:
I'm kind of responsible for this breaking change.

I would help to set it back. PTAL https://github.com/open-policy-agent/opa-envoy-plugin/pull/668

Reply:
That's fair! Let's have this merged and follow up with additions based off of this.

Reply:
Also, if you're interested in reducing allocations, you'll find a lot to like in OPA v1.2.0 🙂